src/java/org/apache/cassandra/db/Mutation.java

```diff
@@ -448,7 +451,7 @@ private Serialization serialization(Mutation mutation, int version)
                     try (DataOutputBuffer dob = DataOutputBuffer.scratchBuffer.get())
                     {
                         serializeInternal(PartitionUpdate.serializer, mutation, dob, version);
-                        serialization = new CachedSerialization(dob.toByteArray());
+                        serialization = new CachedSerialization(dob.unsafeToByteArray());
```

Comment:
if using on heap allocation can we also avoid byte array allocation here with buffer.array()? (buffer.hasArray()). Perhaps provide implementation of CachedSerialization that can use the backed array with offset/length. For larger mutations (ie 64kb-1mb) this can be pretty expensive still

Reply:
err never mind, since using the scratch buffer can't use backing array

Reply:
yes, it is reusable, so we have to copy the result

I have one non-implemented idea for this area of logic - initially we invoke serializedSize and then if the result is not too big we invoke the real serialization. In typical cases mutations are small and we can use some cheaper heuristic to decide if it is small to go directy to  real serialization..

Reply:
If I find a cheap way to do it - I think I will create a separate improvement for it, it is more about CPU saving (while allocations are also affected a bit)